File: go/core/cgroup_reader_test.go
// cgroup v2 file readers — tests mirror python/tests/test_cgroup_reader.py.
//
// Fail-silent contract: every reader returns (zero, false) on missing or
// malformed input. Tests swap the package-level cgroupRoot via t.TempDir.

package core

import (
	"os"
	"path/filepath"
	"testing"
)

// withCgroupRoot points cgroupRoot at tmp and restores it on cleanup.
func withCgroupRoot(t *testing.T, tmp string) {
	t.Helper()
	old := cgroupRoot
	cgroupRoot = tmp
	t.Cleanup(func() { cgroupRoot = old })
}

func seedFile(t *testing.T, dir, name, body string) {
	t.Helper()
	if err := os.WriteFile(filepath.Join(dir, name), []byte(body), 0o644); err != nil {
		t.Fatalf("seed %s: %v", name, err)
	}
}

func TestReadCPUStatParsesUsageUsec(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.stat",
		"usage_usec 12345\nuser_usec 6000\nsystem_usec 6345\n"+
			"nr_periods 0\nnr_throttled 0\nthrottled_usec 0\n")
	withCgroupRoot(t, tmp)

	s, ok := ReadCPUStat()
	if !ok {
		t.Fatal("ReadCPUStat returned ok=false")
	}
	if s.UsageUsec != 12345 {
		t.Fatalf("UsageUsec = %d, want 12345", s.UsageUsec)
	}
}

func TestReadCPUMaxWithQuota(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.max", "100000 100000\n")
	withCgroupRoot(t, tmp)

	m, ok := ReadCPUMax()
	if !ok {
		t.Fatal("ReadCPUMax returned ok=false")
	}
	if m.QuotaUS != 100000 || m.PeriodUS != 100000 || m.VCPUCount != 1.0 {
		t.Fatalf("got %+v, want {Quota:100000 Period:100000 VCPUCount:1.0}", m)
	}
}

func TestReadCPUMaxQuotaFraction(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.max", "25000 100000\n")
	withCgroupRoot(t, tmp)

	m, ok := ReadCPUMax()
	if !ok {
		t.Fatal("ReadCPUMax returned ok=false")
	}
	if m.VCPUCount != 0.25 {
		t.Fatalf("VCPUCount = %v, want 0.25", m.VCPUCount)
	}
}

func TestReadCPUMaxUnlimited(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.max", "max 100000\n")
	withCgroupRoot(t, tmp)

	m, ok := ReadCPUMax()
	if !ok {
		t.Fatal("ReadCPUMax returned ok=false")
	}
	if m.QuotaUS != 0 {
		t.Fatalf("QuotaUS = %d, want 0 (unlimited sentinel)", m.QuotaUS)
	}
	if m.VCPUCount <= 0 {
		t.Fatalf("VCPUCount = %v, want > 0 (nproc fallback)", m.VCPUCount)
	}
}

func TestReadMemoryPeak(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "memory.peak", "2147483648\n")
	withCgroupRoot(t, tmp)

	v, ok := ReadMemoryPeak()
	if !ok || v != 2147483648 {
		t.Fatalf("got (%d, %v)", v, ok)
	}
}

func TestReadMemoryMaxFinite(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "memory.max", "1073741824\n")
	withCgroupRoot(t, tmp)

	v, ok := ReadMemoryMax()
	if !ok || v != 1073741824 {
		t.Fatalf("got (%d, %v)", v, ok)
	}
}

func TestReadMemoryMaxUnlimited(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "memory.max", "max\n")
	withCgroupRoot(t, tmp)

	if _, ok := ReadMemoryMax(); ok {
		t.Fatal("expected ok=false for 'max'")
	}
}

func TestReadMemoryCurrent(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "memory.current", "1024\n")
	withCgroupRoot(t, tmp)

	v, ok := ReadMemoryCurrent()
	if !ok || v != 1024 {
		t.Fatalf("got (%d, %v)", v, ok)
	}
}

func TestMissingFilesReturnFalse(t *testing.T) {
	tmp := t.TempDir()
	withCgroupRoot(t, tmp)

	if _, ok := ReadCPUStat(); ok {
		t.Fatal("ReadCPUStat: expected ok=false")
	}
	if _, ok := ReadCPUMax(); ok {
		t.Fatal("ReadCPUMax: expected ok=false")
	}
	if _, ok := ReadMemoryPeak(); ok {
		t.Fatal("ReadMemoryPeak: expected ok=false")
	}
	if _, ok := ReadMemoryMax(); ok {
		t.Fatal("ReadMemoryMax: expected ok=false")
	}
	if _, ok := ReadMemoryCurrent(); ok {
		t.Fatal("ReadMemoryCurrent: expected ok=false")
	}
}

func TestMalformedCPUStatReturnsFalse(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.stat", "garbage\n")
	withCgroupRoot(t, tmp)

	if _, ok := ReadCPUStat(); ok {
		t.Fatal("expected ok=false on malformed cpu.stat")
	}
}

func TestMalformedCPUMaxReturnsFalse(t *testing.T) {
	tmp := t.TempDir()
	seedFile(t, tmp, "cpu.max", "only-one-token\n")
	withCgroupRoot(t, tmp)

	if _, ok := ReadCPUMax(); ok {
		t.Fatal("expected ok=false on malformed cpu.max")
	}
}

func TestMemoryPeakAbsentWhenKernelTooOld(t *testing.T) {
	// Kernel < 5.19 — memory.peak missing; memory.current present.
	// Reader does NOT fabricate; caller decides fallback (capture spec §6 case 6).
	tmp := t.TempDir()
	seedFile(t, tmp, "memory.current", "1024\n")
	withCgroupRoot(t, tmp)

	if _, ok := ReadMemoryPeak(); ok {
		t.Fatal("expected ok=false when memory.peak file absent")
	}
	v, ok := ReadMemoryCurrent()
	if !ok || v != 1024 {
		t.Fatalf("memory.current got (%d, %v)", v, ok)
	}
}

File: typescript/tests/cgroup-reader.test.ts
/**
 * cgroup v2 file parsing — cpu.stat / cpu.max / memory.peak / memory.max /
 * memory.current.
 *
 * Ports python/tests/test_cgroup_reader.py (12 cases) to vitest.
 */

import { afterEach, beforeEach, describe, expect, test } from "vitest";
import { mkdtempSync, writeFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

import {
  readCpuStat,
  readCpuMax,
  readMemoryPeak,
  readMemoryMax,
  readMemoryCurrent,
  _setCgroupRootForTests,
} from "../src/core/cgroup-reader.js";

let tmp: string;

function seed(files: Record<string, string>): void {
  for (const [name, body] of Object.entries(files)) {
    writeFileSync(join(tmp, name), body);
  }
}

beforeEach(() => {
  tmp = mkdtempSync(join(tmpdir(), "dexcost-cgroup-"));
  _setCgroupRootForTests(tmp);
});

afterEach(() => {
  rmSync(tmp, { recursive: true, force: true });
  _setCgroupRootForTests(null);
});

describe("cgroup reader", () => {
  test("read_cpu_stat parses usage_usec", () => {
    seed({
      "cpu.stat":
        "usage_usec 12345\n" +
        "user_usec 6000\n" +
        "system_usec 6345\n" +
        "nr_periods 0\nnr_throttled 0\nthrottled_usec 0\n",
    });
    const s = readCpuStat();
    expect(s).not.toBeNull();
    expect(s!.usageUsec).toBe(12345);
  });

  test("read_cpu_max with quota", () => {
    seed({ "cpu.max": "100000 100000\n" });
    const m = readCpuMax();
    expect(m).toEqual({ quotaUs: 100000, periodUs: 100000, vcpuCount: 1.0 });
  });

  test("read_cpu_max quota fraction", () => {
    // 25000 / 100000 = 0.25 vCPU (a small Fargate task).
    seed({ "cpu.max": "25000 100000\n" });
    const m = readCpuMax();
    expect(m).not.toBeNull();
    expect(m!.vcpuCount).toBe(0.25);
  });

  test("read_cpu_max unlimited falls back to nproc", () => {
    seed({ "cpu.max": "max 100000\n" });
    const m = readCpuMax();
    expect(m).not.toBeNull();
    expect(m!.quotaUs).toBeNull();
    expect(m!.vcpuCount).toBeGreaterThan(0);
  });

  test("read_memory_peak", () => {
    seed({ "memory.peak": "2147483648\n" });
    expect(readMemoryPeak()).toBe(2147483648);
  });

  test("read_memory_max finite", () => {
    seed({ "memory.max": "1073741824\n" });
    expect(readMemoryMax()).toBe(1073741824);
  });

  test("read_memory_max unlimited returns null", () => {
    seed({ "memory.max": "max\n" });
    expect(readMemoryMax()).toBeNull();
  });

  test("read_memory_current", () => {
    seed({ "memory.current": "1024\n" });
    expect(readMemoryCurrent()).toBe(1024);
  });

  test("missing files return null", () => {
    expect(readCpuStat()).toBeNull();
    expect(readCpuMax()).toBeNull();
    expect(readMemoryPeak()).toBeNull();
    expect(readMemoryMax()).toBeNull();
    expect(readMemoryCurrent()).toBeNull();
  });

  test("malformed cpu.stat returns null", () => {
    seed({ "cpu.stat": "garbage\n" });
    expect(readCpuStat()).toBeNull();
  });

  test("malformed cpu.max returns null", () => {
    seed({ "cpu.max": "only-one-token\n" });
    expect(readCpuMax()).toBeNull();
  });

  test("memory.peak absent when kernel too old, current still present", () => {
    seed({ "memory.current": "1024\n" });
    expect(readMemoryPeak()).toBeNull();
    expect(readMemoryCurrent()).toBe(1024);
  });
});
